stock_manager/controllers/abstract.py

```diff
@@ -0,0 +1,233 @@
+from abc import ABC, ABCMeta, abstractmethod
+from pathlib import Path
+from typing import override, TYPE_CHECKING
+
+from numpy import ndarray
+from PyQt6.QtCore import pyqtSignal, QThread
+from PyQt6.QtWidgets import QMessageBox, QTableWidget, QTableWidgetItem, QWidget
+from PyQt6.uic import loadUi
+
+import stock_manager
+
+if TYPE_CHECKING:
+	from stock_manager import App
+	from stock_manager import Item
+
+
+class CombinedMeta(type(QWidget), ABCMeta):
+	"""
+	A metaclass combining PyQts QWidget metaclass and Python's ABCMeta.
+	
+	This enables the AbstractController to inherit from both PyQt widgets and Python abstract base classes,
+	resolving metaclass conflicts that would otherwise occur with multiple inheritance.
+	"""
+	pass
+
+
+class AbstractController(ABC, QWidget, metaclass=CombinedMeta):
+	"""Abstract controller with common UI behavior for the application."""
+	
+	def __init__(self, file_name: str, app: 'App'):
+		"""
+		Initialize the abstract controller, load its UI, and set up sidebar navigation handlers.
+		
+		:param file_name: The name of the .ui file (without extension) to load for this controller.
+		:param app: Reference to the main application instance
+		"""
+		
+		super(AbstractController, self).__init__()
+		
+		self.app = app
+		self.logger = app.log
+		self.database = app.db
+		self.PAGE_INDEX = -1
+		self.PAGE_NAME: stock_manager.PageNames | None = None
+		
+		try:
+			ui_path = Path(__file__).resolve().parent.parent.parent / 'ui' / f'{file_name}.ui'
+			loadUi(str(ui_path), self)
+		except Exception as e:
+			print(f'Failed To Load {file_name}.ui File: {e}')
+			self.logger.error_log(f"Failed To Load {file_name}.ui File: {e}")
+			QMessageBox.critical(
+					self,
+					f'{file_name}.ui Failure',
+					f'Failed To Load {file_name}.ui File',
+					QMessageBox.StandardButton.Ok
+			)
+	
+	@abstractmethod
+	def handle_connections(self) -> None:
+		pass
+	
+	@staticmethod
+	def filter_table(text: str, table: QTableWidget) -> None:
```

Comment:
Qt has a FilterProxyModel for handling behavior like this.  It's a bit more complicated to set up, but it is a bit more robust than manually showing / hiding rows
https://doc.qt.io/qt-6/qsortfilterproxymodel.html